test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/MaxInputBufferLengthTests.cs

```diff
@@ -0,0 +1,250 @@
+                    // On Windows, the client is usually paused at (MaxInputBufferSize + 700,000).
+                    Tuple.Create((int?)5 * 1024 * 1024, true),
+                    
+                    // Even though maxInputBufferLength < _dataLength, client should not be paused since the
+                    // OS-level buffers in client and/or server will handle the overflow.
+                    Tuple.Create((int?)10 * 1024 * 1024 - 1, false),
+
+                    // Buffer is exactly the same size as data.  Exposed race condition where
+                    // IConnectionControl.Resume() was called after socket was disconnected.
+                    Tuple.Create((int?)10 * 1024 * 1024, false),
+
+                    // Largest possible buffer, should never trigger backpressure.
+                    Tuple.Create((int?)Int32.MaxValue, false),
+
+                    // Disables all code related to computing and limiting the size of the input buffer.
+                    Tuple.Create((int?)null, false)
+                };
+                var sendContentLengthHeaderValues = new[] { true, false };
+                var sslValues = new[] { true, false };
+
+                return from maxInputBufferLength in maxInputBufferLengthValues
+                       from sendContentLengthHeader in sendContentLengthHeaderValues
+                       from ssl in sslValues
+                       select new object[] {
+                           maxInputBufferLength.Item1,
+                           sendContentLengthHeader,
+                           ssl,
+                           maxInputBufferLength.Item2
+                       };
+            }
+        }
+
+        [Theory]
+        [MemberData("LargeUploadData")]
+        public async Task LargeUpload(int? maxInputBufferLength, bool sendContentLengthHeader, bool ssl, bool expectPause)
+        {
+            // Parameters
+            var data = new byte[_dataLength];
+            var bytesWrittenTimeout = TimeSpan.FromMilliseconds(100);
+            var maxSendSize = 4096;
+
+            // Initialize data with random bytes
+            (new Random()).NextBytes(data);
+
+            var startReadingRequestBody = new ManualResetEvent(false);
+            var clientFinishedSendingRequestBody = new ManualResetEvent(false);
+            var bytesWrittenEvent = new AutoResetEvent(false);
+
+            using (var host = StartWebHost(maxInputBufferLength, data, startReadingRequestBody, clientFinishedSendingRequestBody))
+            {
+                var port = host.GetPort(ssl ? "https" : "http");
+                using (var socket = CreateSocket(port))
+                using (var stream = await CreateStreamAsync(socket, ssl, host.GetHost()))
+                {
+                    await WritePostRequestHeaders(stream, sendContentLengthHeader ? (int?)data.Length : null);
+
+                    var bytesWritten = 0;
+
+                    Func<Task> sendFunc = async () =>
```

Comment:
Which part?  I wanted to switch from sync to async writes, in part to improve test performance, and also because async IO is generally better and what we recommend customers use.  This requires creating an async delegate.

This can then be invoked either directly or by calling `Task.Run(Func<Task>)`, but I don't see any benefit to `Task.Run()`.